File: Python3/494. Target Sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        
        def dfs(nums, i, sums, memo):
            #print(i, sums, memo)
            
            if (i, sums) in memo:
                return memo[(i, sums)]
            
            # reaches the end of the nums array
            # the check condition is i == len(nums) because
            # we perform dfs(.., i+1, ..) before 
            if i == len(nums):
                return target == sums
            
            count = 0
            count += dfs(nums, i+1, sums+nums[i], memo)
            count += dfs(nums, i+1, sums-nums[i], memo)
            memo[(i, sums)] = count
            return count
        
        memo = {}
        
        return dfs(nums, 0, 0, memo)
        
        

File: Java/494. Target Sum.java
class Solution {
    
    public int findTargetSumWays(int[] nums, int target) {
        
        return dfs(nums, 0, target, new HashMap<>());
    }
    
    private int dfs(int[] nums, int i, int sum, Map<String, Integer> memo){
        String temp = i+"->"+sum;
        // System.out.println(temp);
        
        if(memo.containsKey(temp)){
            // System.out.println(temp + "\t" + memo.get(temp));
            return memo.get(temp);
            
        }
        
        if(i == nums.length){
            return sum == 0 ? 1 : 0;
        }
        
        int count = 0;
        count += dfs(nums, i+1, sum-nums[i], memo);
        count += dfs(nums, i+1, sum+nums[i], memo);
        memo.put(temp, count);
        return count;
    }
}